File: backend/app/services/payments/provider_base.py
from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass
from datetime import datetime
from decimal import Decimal
from typing import Any, Dict, Optional


@dataclass
class CreateIntentResult:
    provider: str
    merchant_txn_id: str
    checkout_url: Optional[str]
    upi_intent_url: Optional[str]
    qr_payload: Optional[str]
    expires_at: Optional[datetime]
    gateway_txn_id: Optional[str]
    raw_response: Dict[str, Any]


@dataclass
class WebhookVerificationResult:
    merchant_txn_id: str
    gateway_txn_id: Optional[str]
    status: str
    amount: Optional[Decimal]
    raw_event: Dict[str, Any]


@dataclass
class FetchStatusResult:
    status: str
    gateway_txn_id: Optional[str]
    raw_status: Dict[str, Any]


class PaymentProvider(ABC):
    provider_name: str

    def __init__(self, config: Dict[str, Any]):
        self.config = config

    @abstractmethod
    def create_intent(
        self,
        *,
        amount: Decimal,
        merchant_txn_id: str,
        student_name: str,
        student_email: Optional[str],
        student_phone: Optional[str],
        metadata: Dict[str, Any],
    ) -> CreateIntentResult:
        raise NotImplementedError

    @abstractmethod
    def verify_webhook(self, headers: Dict[str, str], body: bytes) -> WebhookVerificationResult:
        raise NotImplementedError

    @abstractmethod
    def fetch_status(self, *, merchant_txn_id: str, gateway_txn_id: Optional[str] = None) -> FetchStatusResult:
        raise NotImplementedError

File: backend-next/lib/services/payments/provider-base.ts
export interface CreateIntentResult {
  provider: string;
  merchant_txn_id: string;
  checkout_url?: string | null;
  upi_intent_url?: string | null;
  qr_payload?: string | null;
  expires_at?: Date | null;
  gateway_txn_id?: string | null;
  raw_response: any;
}

export interface WebhookVerificationResult {
  merchant_txn_id: string;
  gateway_txn_id?: string | null;
  status: "SUCCESS" | "FAILED" | "PENDING" | "CANCELLED" | "EXPIRED";
  amount?: number | null;
  raw_event: any;
}

export interface FetchStatusResult {
  status: "SUCCESS" | "FAILED" | "PENDING" | "CANCELLED" | "EXPIRED";
  gateway_txn_id?: string | null;
  raw_status: any;
}

export abstract class PaymentProvider {
  constructor(protected config: any) {}
  
  abstract createIntent(data: {
    amount: number;
    merchant_txn_id: string;
    tenant_name: string;
    tenant_email?: string;
    tenant_phone?: string;
    metadata: any;
  }): Promise<CreateIntentResult>;

  abstract verifyWebhook(headers: Record<string, string>, body: string | Buffer): Promise<WebhookVerificationResult>;

  abstract fetchStatus(merchant_txn_id: string, gateway_txn_id?: string): Promise<FetchStatusResult>;
}
